velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,226 @@
+  // delimiter is empty, the result does not include an empty tail string, e.g.
+  // split('abc', '') outputs ["a", "b", "c"] instead of ["a", "b", "c", ""].
+  // The result does not include remaining string when limit is smaller than the
+  // string size, e.g. split('abc', '', 2) outputs ["a", "b"] instead of ["a",
+  // "bc"].
+  void splitEmptyDelimiter(
+      const StringView current,
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+    if (current.size() == 0) {
+      arrayWriter.add_item().setNoCopy(StringView());
+      resultWriter.commit();
+      return;
+    }
+
+    const char* const begin = current.begin();
+    const char* const end = current.end();
+    const char* pos = begin;
+    while (pos < end && pos < limit + begin) {
+      arrayWriter.add_item().setNoCopy(StringView(pos, 1));
+      pos += 1;
+    }
+    resultWriter.commit();
+  }
 
-  void apply(
-      const SelectivityVector& rows,
-      std::vector<VectorPtr>& args,
-      const TypePtr& /* outputType */,
-      exec::EvalCtx& context,
-      VectorPtr& result) const override {
-    auto delimiterVector = args[1]->as<ConstantVector<StringView>>();
-    VELOX_CHECK(
-        delimiterVector, "Split function supports only constant delimiter");
-    auto patternString = args[1]->as<ConstantVector<StringView>>()->valueAt(0);
-    VELOX_CHECK_EQ(
-        patternString.size(),
-        1,
-        "split only supports only single-character pattern");
-    char pattern = patternString.data()[0];
-    SplitCharacter splitCharacter(pattern);
-    splitCharacter.apply(rows, args, nullptr, context, result);
+  // Split with a non-empty pattern.
+  void splitInner(
+      StringView input,
+      const StringView delim,
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    VELOX_CHECK(!delim.empty(), "Delimiter size should not empty");
+    // Add new array (for the new row) to our array vector.
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+
+    // Trivial case of converting string to array with 1 element.
+    if (limit == 1) {
+      arrayWriter.add_item().setNoCopy(input);
+      resultWriter.commit();
+      return;
+    }
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // add them to the elements vector, until we reach the end of the
+    // string or the limit.
+    int32_t addedElements{0};
+    auto* re = cache_.findOrCompile(delim);
+    const auto re2String = re2::StringPiece(input.data(), input.size());
+    size_t pos = 0;
+    const char* start = input.data();
+    re2::StringPiece subMatches[1];
+    while (re->Match(
+        re2String, pos, input.size(), RE2::Anchor::UNANCHORED, subMatches, 1)) {
```

Comment:
Please explain how we match re2String against delim. Some example:

> Matches a regular expression against a portion of the input string, starting from 'pos' to the end of the input string. The match is not anchored, which means it can start at any position in the string. If a match is found, the matched portion of the string is stored in 'subMatches'. The '1' indicates that we are only interested in the first match found.